## Тема 1. Первичный анализ данных

### Agenda
В данном разделе вы познакомитесь с основными библиотеками python для анализа данных: pandas, matplotlib и попробуете на практике провести анализ и сделать выводы по исходным данным.
В качестве исходных данных был выбран датасет титаника (https://www.kaggle.com/c/titanic/data).

### Jupyter notebook
В качестве среды разработки используется Jupyter notebook. Основными достоинствами которого являются:
* Возможность выполнения кода ячейками
* Генерация выходного файла в виде python кода, pdf или html страницы
* Отображение графиков прямо внутри тетради
* Возможность использовать LaTeX, Markdown и HTML разметку для большей наглядности

### Основные операции в Jupyter notebook
Ячейка - базовая еденица jupyter

### Описание датасета "Титаник"

Потопление титаника является одним из самых ужасных крушений в истории. 15 апреля 1912 во время своей первого рейса, титаник столкнулся с айсбергом и затонул, что повлекло смерти 1502 человек из 2224 пассажиров и комманды. Эта сенсационная трагедия шокировала международное сообщество и повлекло за собой более чщательное регулирование судостроительства.

The sinking of the RMS Titanic is one of the most infamous shipwrecks in history.  On April 15, 1912, during her maiden voyage, the Titanic sank after colliding with an iceberg, killing 1502 out of 2224 passengers and crew. This sensational tragedy shocked the international community and led to better safety regulations for ships.

Одна из причин, почему крушение повлекло столько потерь заключалась, в том, что не было в достаточном количество спасательных шлюпок как для пассажиров, так и для команды. Так или иначе в выживании сыграла роль не только удача, неготорые группам людей больше удалось выжить, чем остальным, к таким группам относятся женщины, дети, и пассажиры купившие билет первого класса.

One of the reasons that the shipwreck led to such loss of life was that there were not enough lifeboats for the passengers and crew. Although there was some element of luck involved in surviving the sinking, some groups of people were more likely to survive than others, such as women, children, and the upper-class.

Набор данных содержит записи о пассажиров титаника и состоит из следующих столбцов:

| Столбец   |      Определение      |  Значения |
|----------|---------|------:|
| survival | Выжил | 	0 = Нет, 1 = Да
| pclass   | Класс билета  |	1 = Первый, 2 = Второй, 3 = Третий
| sex 	   | Пол | male = Мужской, female = Женский 	
| Age 	   | Возраст в годах | Количественный тип
| sibsp    | Количество братьев и сестер | Количественный тип
| parch    | Количество родителей и детей |	Количественный тип
| ticket   | Номер билета 	| Количественный тип
| fare 	   | Цена	| Количественный тип
| cabin    | Кабина |	
| embarked | Место посадки |

### Выполнение лабораторной работы

In [2]:
# Загрузим необходимы библиотеки
import pandas as pd
pd.set_option('display.max.columns', 100)
import matplotlib.pyplot as plt
import seaborn as sns

# Для возможности вывода картинок прямо в тетрадь
%matplotlib inline 

In [18]:
# Загрузка дата сета
data = pd.read_csv("datasets/titanic_train.csv", index_col="PassengerId")

# Вывод на экран первых 5 записей
data.head()

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   

             SibSp  Parch            Ticket     Fare Cabin Embarked  
PassengerId                                                          
1                1      0         A/5 21171   7.2500   NaN        S  
2                1      0          PC 17599  71.2833   C85        C  
3                0      0  STON/O2. 3101282   7.9250   NaN        S  
4                1      0            113803  53.1000  C123        S  
5                0      0            373450   8.0500   NaN        S

При просмотре исходных данных очень удобно просмотреть несколько как записей начиная с начала, так и несколько записей начиная с конца. В pandas для этого используются такие методы как **head(n)**, в который передается количество записей и результатом данной функции будет первые n записей. **tail(n)** позволяет получить n последний записей.

Для получеия значений отдельного столбца, его название записывается в фигурных скобках - **dataset["columnName"]**, либо можно обратится через точку - **dataset.columnName**

Основные операции с данными в pandas представлены в документации на данной странице - https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.html

In [22]:
# Выведем имена первых 10 пасажиров
data["Name"].head(10)

PassengerId
1                               Braund, Mr. Owen Harris
2     Cumings, Mrs. John Bradley (Florence Briggs Th...
3                                Heikkinen, Miss. Laina
4          Futrelle, Mrs. Jacques Heath (Lily May Peel)
5                              Allen, Mr. William Henry
6                                      Moran, Mr. James
7                               McCarthy, Mr. Timothy J
8                        Palsson, Master. Gosta Leonard
9     Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)
10                  Nasser, Mrs. Nicholas (Adele Achem)
Name: Name, dtype: object

In [23]:
# Для обращения к нескольким столбцам
# необходимо передать список из названий стобцов
data[["Name", "Age"]].head(10)

Name   Age
PassengerId                                                         
1                                      Braund, Mr. Owen Harris  22.0
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  38.0
3                                       Heikkinen, Miss. Laina  26.0
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  35.0
5                                     Allen, Mr. William Henry  35.0
6                                             Moran, Mr. James   NaN
7                                      McCarthy, Mr. Timothy J  54.0
8                               Palsson, Master. Gosta Leonard   2.0
9            Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  27.0
10                         Nasser, Mrs. Nicholas (Adele Achem)  14.0

In [24]:
# Получим данные о размерности
lines, cols = data.shape
print("Количество строк = {}\nКоличество стоблцов = {}".format(lines, cols))

# Выведем названия стобцов
print(data.columns)

Количество строк = 891
Количество стоблцов = 11
Index(['Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch', 'Ticket',
       'Fare', 'Cabin', 'Embarked'],
      dtype='object')


In [12]:
# Получим информацию о датасете и типах внутри него
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
PassengerId    891 non-null int64
Survived       891 non-null int64
Pclass         891 non-null int64
Name           891 non-null object
Sex            891 non-null object
Age            714 non-null float64
SibSp          891 non-null int64
Parch          891 non-null int64
Ticket         891 non-null object
Fare           891 non-null float64
Cabin          204 non-null object
Embarked       889 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 66.2+ KB


Как видно из данных, датасет содержит 891 запись, но далеко не все ячейки содержат данных. Поле возраста (Age) заполнено только для 714 записей, а поле занятой кабины (Cabin) только для 204. 
Существует множество различных способов как можно поступить в данной ситуации:
* Удалить строки с пропусками
* Заполнить пропуски средним значением или модой (наиболее часто встречающимся значением)
* и тд


In [48]:
# Вывод количества спасшихся пассажиров
data_full["Survived"].value_counts()

0    549
1    340
Name: Survived, dtype: int64

In [49]:
# Вывод количества людей севших на корабль в одном из портов
data_full["Embarked"].value_counts()

S    644
C    168
Q     77
Name: Embarked, dtype: int64

При анализе данных бывает учень удобно сравнить несколько стоблцов между собой. Для таких случаев существуют сводные таблицы. Сводные таблицы реализует метод **crosstab**, в который необходимо передать исследуемые стоблцы.

In [54]:
# Создание сводной таблицы между полом и выживаемостью
pd.crosstab(data["Sex"], data["Survived"])

Survived    0    1
Sex               
female     81  233
male      468  109

Как видно большинству женщин удалось спастить, а мужчины в основном погибли. Эта информаци яможет быть полезной при построении предсказательной модели выживаемости того или иного пассажира.

Как вы помните исходные данные содержат пропуски. Для использования данных на предсказательных моделях необходимо избавится от пропусков.

Поле возраста можно заполнить средними значениями по всем данным. Для этого используем функцию **fillna(value)**, которая заполняет пустые значения значением value. Для вычисления среднего значения по всему возрасту применим функцию **mean()**

Столбец Embarked не заполнен только в 2-х местах, поэтому эти записи можно без сильных потерь для информативности датасета удалить. 

In [40]:
# Дополним пустые ячейки средним значением
data["Age"] = data["Age"].fillna(data["Age"].mean())

# Создадим дубликат датасета без стоблца Cabin и удалим записи содержащие пустые ячейки
data_full = data.drop("Cabin", axis=1).dropna()

In [41]:
# Проверим новый датасет
data_full.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 889 entries, 1 to 891
Data columns (total 10 columns):
Survived    889 non-null int64
Pclass      889 non-null int64
Name        889 non-null object
Sex         889 non-null object
Age         889 non-null float64
SibSp       889 non-null int64
Parch       889 non-null int64
Ticket      889 non-null object
Fare        889 non-null float64
Embarked    889 non-null object
dtypes: float64(2), int64(4), object(4)
memory usage: 62.5+ KB


Для применения данных в моделях машинного обучения их необходимо привести к числовому виду. К примеру, в нашем датасете пол обозначен строковым типом данных, их необходимо преобразовать в числовой. С этой задачей отлично справляется метод **map**. В данный метод необходимо передать словарь замен и в результате будет получен вектор с заменеными данными.

In [58]:
# Заменяем строку "male" на 0, "female" на 1
data_full["gender"] = data_full["Sex"].map({"male": 0, "female": 1})
data_full["gender"].head()

PassengerId
1    0
2    1
3    1
4    1
5    0
Name: gender, dtype: int64